File: Python/SampleProperties.py
# -*- coding: utf-8 -*-
"""
Create Time: 2021/1/29 8:36
Author: Bo Wang
"""
import numpy as np
from numpy.fft import fft
from FrequencyAnalysisUtil import concatenateTimeSeries, calRangeOfFrequency
from scipy import signal
#此文件由王博开发，主要实现吸收系数与折射率的计算
class SampleProperties:
    """
    This class is to calculate the refractive index and the absorption rate of 
    a subject scanned by the THz-TDS. An initial measurement without loading the 
    sample has to be made as reference signal.

    Attributes:
        refT: the time points of the reference signal
        refX: the amplitudes of the reference signal
        sampT: the time point of the sample signal
        sampX: the amplitudes of the sample signal
        d: the depth of the sample
        mode: the way to fit the phase change and the frequency. 'linear_fit', use linear regression to fit the phase change.
        'phase_comp', subtract the error induced by the anterior portion of the signal from the phase change. The default
        is 'none'.
        lowLimit: the lower limit of the partition of high SNR for the linear fit and the phase compensation.
        upLimit: the upper limit of the partition of high SNR for the linear fit and the phase compensation.
        denoise: if true, remove the reflection peaks by deconvolution
        f(np.array): the frequency points of the absorption rate and refractive index
    
    @author: Bo Wang
    """

    def __init__(self, refT, refX, sampT, sampX, d, mode = 'none', lowLimit = 0.3, upLimit = 0.5, denoise = False):
        if mode not in ['linear_fit', 'phase_comp', 'none']:
            raise ValueError("The mode should be linear_fit, phase_comp")

        self.refT = refT[:]
        self.refX = refX[:]
        self.sampT = sampT[:]
        self.sampX = sampX[:]
        self.d = d
        self.mode = mode
        self.lowLimit = lowLimit
        self.upLimit = upLimit
        self.denoise = denoise
        self.__calTransFunction()


    def calRefractiveIndex(self):
        """
        Calculate the sample's refractive index.

        Returns:
            refractiveIndex(np.array): a 1d np array of type float64
        """
        # The light speed
        c = 3e8
        refractiveIndex = self.__phaseTransFunction * c
        refractiveIndex /= (2 * np.pi * 1e9)
        refractiveIndex /= self.f
        refractiveIndex /= self.d
        refractiveIndex += 1

        # refractiveIndex = self.DealRefractive(refractiveIndex)

        return refractiveIndex

    def calExtinction(self):
        """
        Calculate the sample's absorption rate.

        Returns:
            absorptionRate(np.array): a 1d np array of float64 representing the absorption rates
        """

        # refractiveIndex = self.__linearFitRefractIndex()
        refractiveIndex = self.calRefractiveIndex()
        refractiveIndex = np.mean(refractiveIndex[(self.f > 0.6) & (self.f < 1.6)])
        absorptionRate = 4 * refractiveIndex
        absorptionRate /= self.__ampTransFunction
        absorptionRate /= np.power(refractiveIndex + 1, 2)
        if np.min(absorptionRate) < 0:
            absorptionRate = absorptionRate - np.min(absorptionRate) + 1
        absorptionRate = np.log(absorptionRate)

        # Convert the unit to cm^-1
        absorptionRate = absorptionRate*0.3/(self.d*self.f)

        return absorptionRate

    def calAbsorptionRate(self):
        """
        Calculate the sample's absorption rate.

        Returns:
            absorptionRate(np.array): a 1d np array of float64 representing the absorption rates
        """

        refractiveIndex = self.calRefractiveIndex()
        refractiveIndex = np.mean(refractiveIndex[(self.f > 0.6) & (self.f < 1.6)])
        absorptionRate = 4 * refractiveIndex
        absorptionRate /= self.__ampTransFunction
        absorptionRate /= np.power(refractiveIndex + 1, 2)
        if np.min(absorptionRate) < 0:
            absorptionRate = absorptionRate - np.min(absorptionRate) + 1
        absorptionRate = np.log(absorptionRate)

        # Convert the unit to cm^-1
        absorptionRate *= 2 / (self.d * 0.1)

        return absorptionRate

    #####################################Private Methods#######################

    def __calPhaseAndAmplitude(self, t, refX, sampX, indices):
        """
        Calculate the phase and amplitude of a frequency spectrum by
        converting the time spectrum

        Args:
            refX: the time spectrum of the reference signal
            sampX: the time spectrum of the sample signal
            indices: the indices representing the effective range
        """

        refXf = fft(refX)
        sampXf = fft(sampX)

        xf = sampXf / refXf
        amplitudes = np.abs(xf)
        phase = np.angle(xf)

        return phase[indices], amplitudes[indices]

    def __calTransFunction(self):
        """
        Calculate the sample's transmission function using the time spectrum
        of the sample and the reference signal
        """
        # The default length of the time signal 
        defDuration = 100.0
        sampDuration = self.sampT[-1] - self.sampT[0]
        refDuration = self.refT[-1] - self.refT[0]

        # Concatenate the time series to the default duration or the longer duration 
        # of the reference series and the sample series
        duration = max(defDuration, sampDuration, refDuration)
        newSampT, newSampX = concatenateTimeSeries(self.sampT, self.sampX, duration)
        self.sampX = newSampX
        _, newRefX = concatenateTimeSeries(self.refT, self.refX, duration)
        self.refX = newRefX
        f, _ = calRangeOfFrequency(newSampT, len(newSampT), False)
        f = np.array(f)

        # The frequency spectrum below 0.1 THz may contain peculiar values, therefore remove the portion below 0.1 THz
        effectLowLimit = 0.1
        indices = f >= effectLowLimit
        self.f = f[indices]
        self.refraction_ratio = np.abs(np.fft.fft(self.refX))/np.abs(np.fft.fft(self.sampX))
        self.refraction_ratio = self.refraction_ratio[indices]


        phaseTrans, self.__ampTransFunction = self.__calPhaseAndAmplitude(newSampT, newRefX, newSampX, indices)

        # Correct the jumps between consecutive phase shifts to be within pi.
        phaseTrans = -np.unwrap(phaseTrans)
        f1 = np.argmin(np.abs(f - 0.2))
        f2 = np.argmin(np.abs(f - 0.6))

        t = range(len(f[f1:f2])) + f1
        params = np.polyfit(t, phaseTrans[f1:f2], 1)
        phaseTrans = phaseTrans - params[1]

        if self.mode == 'phase_comp':
            indices, theorPhase = self.__calTheoreticalPhase()
            phaseOffset = np.mean(phaseTrans[indices] - theorPhase) / np.pi
            phaseOffset = np.round(phaseOffset) * np.pi
            self.__phaseTransFunction = phaseTrans - phaseOffset
        else:
            self.__phaseTransFunction = phaseTrans

    def __calculateSlope(self, phase, f):
        """ Use y = x * k to estimate k, thereby trans(x) * y = trans(x) * x * k, to assure that the fitted line goes
        through the origin."""
        phase = np.array(phase)
        f = np.array(f)
        x = np.sum(np.power(phase,2))
        y = np.sum(np.multiply(phase, f))
        k = y / x
        return k

    def __linearFitRefractIndex(self):
        """Make linear fitting to the sample's refractive index

        Args:
            see calAbsorptionRate

        Returns:
            a float representing the corrected refractive index
        """
        # Fit the phase shift
        indicef = np.array(np.where((self.f >= self.lowLimit) & (self.f <= self.upLimit)))
        indicef = indicef.reshape(indicef.shape[1],)
        indices = indicef.tolist()

        f = self.f[indices]
        phaseTrans = self.__phaseTransFunction[indices]

        # Make linear fitting
        slope = self.__calculateSlope(phaseTrans, f)

        # The light speed
        c = 3e8
        factor = c / self.d / (2 * np.pi * 1e9)
        refractiveIndex = factor * slope + 1

        return refractiveIndex

    def __calTheoreticalPhase(self):
        """ Calculate the theoretical phase assuming no dispersion occurs in the sample
        """
        indicef = np.array(np.where((self.f >= self.lowLimit) & (self.f <= self.upLimit)))
        indicef = indicef.reshape(indicef.shape[1],)
        indices = indicef.tolist()

        f = self.f[indices]
        # f = self.f[(self.f >= self.lowLimit) & (self.f <= self.upLimit)]
        delay = self.__calTimeDelay()
        theorPhase = 2 * np.pi * f * delay

        return indices, theorPhase

    def __calTimeDelay(self):
        """Calculate the time delay of the primary peak of the sample signal by comparing with the reference signal."""
        refMaxIndex = np.argmax(np.array(self.refX))
        refPeakTime = self.refT[refMaxIndex]
        sampMaxIndex = np.argmax(np.array(self.sampX))
        sampPeakTime = self.sampT[sampMaxIndex]
        delay = sampPeakTime - refPeakTime

        return delay

File: Matlab/SampleProperties.m
% This class is to calculate the refractive index and the absorption rate
% of a sample measured by the THz system. An initial measurement without the  
% sample has to be applied before the ones with the sample. 
classdef SampleProperties
    properties
        % The discrete times of the reference time-domain spectrum
        refT
        % The amplitudes of the reference time-domain spectrum
        refX
        % The discrete times of the time-domain spectrum of the sample
        sampT
        % The amplitudes of the time-domain spectrum of the sample
        sampX
        % The depth of the sample, unit mm, usually between 1 to 2 mm
        d{mustBeNumeric}
        % If true compensate the phase transition function by the
        % difference between the estimated phase transition and that with
        % no dispersion
        compPhase
        % If true, use the deconvolutional denooising to eliminate the
        % reflection peaks
        denoise
        % The ratio of the frequency amplitude of the sample and the
        % frequency amplitude of the empty mining
        ampTransFunction
        % The difference between the phase of the frequency spectrum of the
        % sample and that of the empty mining
        phaseTransFunction
        % The discrete frequencies of the smaple's frequency-domain
        % spectrum, unit THz
        f
    end
    methods
        function obj = SampleProperties(refT, refX, sampT, sampX, d, compPhase, denoise)
            if nargin == 5
                obj.refT = refT;
                obj.refX = refX;
                obj.sampT = sampT;
                obj.sampX = sampX;
                obj.d = d;
                obj.compPhase = false;
                obj.denoise = false;
            elseif nargin == 6
                obj.refT = refT;
                obj.refX = refX;
                obj.sampT = sampT;
                obj.sampX = sampX;
                obj.d = d;
                obj.compPhase = compPhase;
                obj.denoise = false;
            elseif nargin == 7
                obj.refT = refT;
                obj.refX = refX;
                obj.sampT = sampT;
                obj.sampX = sampX;
                obj.d = d;
                obj.compPhase = compPhase;
                obj.denoise = denoise;
            else
                error('The number of input arguments shoulb be five.')
            end
            
            % Calculate the transmission function
            [obj.ampTransFunction, obj.phaseTransFunction, obj.f] = obj.calTransFunction();
        end
        
        % Calculate the sample's refractive index 
        function sampRefraction = calSampleRefraction(obj)
            % The light spped in vacuum, unit m/s
            c = 3e8;            
            sampRefraction = obj.phaseTransFunction * c;
            sampRefraction = sampRefraction / (2 * pi) ./ obj.f / obj.d / 1e9; 
            sampRefraction = sampRefraction + 1;
        end
        
         % Calculate the smaple's absorption rate
        function sampAbsortption = calSampleAbsorption(obj)
            % Calculate the corrected refraction rate
            n = obj.getCorrectedRefraction();
            
            % Absorptioin rate, cited from 'Lu 2013'
            sampAbsortption = log(4 * n ./ obj.ampTransFunction ./ (1 + n).^2);
            
            % Unit cm^-1
            sampAbsortption = 2 / (obj.d * 0.1) * sampAbsortption; 
        end
    end
    
    methods(Access = private)
        % Get the phase and the amplitude of a frequency spectrum by
        % converting from the time spectrum
        function [phase, amp] = getPhaseAndAmplitude(obj, x, f)
            if obj.denoise
                xf = FrequencyAnalysisUtil.removeReflectionPeaks(obj.refT, f, x);
            else
                xf = fft(x);
            end
            phase = angle(xf);
            amp = abs(xf);
        end
        
        % Make linear fitting to the sample's refraction rate
        function n = getCorrectedRefraction(obj)
            % Fit the phase transition from 0.3 THz to 0.5 THz
            lowLimit = 0.3;
            upLimit = 0.5;
            range = obj.f > lowLimit & obj.f < upLimit;
            slope = obj.phaseTransFunction(range) \ obj.f(range);
            
            % Light speed m/s
            c = 3.0e8;
            
            % c / (d * 2pi)
            factor = c / obj.d / (2 * pi) / 1e9;
            n = factor * slope + 1;
        end 
        
        % Calculate the sample's transmission function using the time-domain
        % spectrum of the system without the sample and that with the
        % sample 
        function [ampTrans, phaseTrans, f] = calTransFunction(obj)
            % The default duration is 100 ps
            duration = 100;
            sampDuration = obj.sampT(end) - obj.sampT(1);
            refDuration = obj.refT(end) - obj.refT(1);

            % Concatenate the time series to the longest duration
            duration = max([duration, sampDuration, refDuration]);
            [newSampT, newSampX] = FrequencyAnalysisUtil.concatenateTimeSeries(obj.sampT, obj.sampX, duration);
            [~, newRefX] = FrequencyAnalysisUtil.concatenateTimeSeries(obj.refT, obj.refX, duration);
            f = FrequencyAnalysisUtil.calRangeOfFrequency(newSampT, length(newSampT), false);
            
%             [sampPhase, sampAmp] = obj.getPhaseAndAmplitude(newSampX, f); 
%             [refPhase, refAmp] = obj.getPhaseAndAmplitude(newRefX, f);
            refXf = fft(newRefX);
            sampXf = fft(newSampX);
            xf = sampXf ./ refXf;
            ampTrans = abs(xf);
            phaseDiff = angle(xf);
%             ampTrans = sampAmp ./ refAmp;
            
            
            
            % The lower limit of the effective signal
            sigLowLimit = 0.1;
            slIndices = f >= sigLowLimit;
%             phaseDiff = sampPhase - refPhase;
            phaseDiff = phaseDiff(slIndices);
            ampTrans = ampTrans(slIndices);
            f = f(slIndices);
            
            % Make the sawtooth-shape angle series continuous
            phaseTrans = -unwrap(phaseDiff);
            
            if obj.compPhase
                % Calculate the offset between the estimated phase and that
                % without dispersion 
                delay = obj.calTimeDelay();
                noDispPhaseTrans = 2 * pi * f * delay;

                % The portion of the highest SNR
                highSnrIndices = f > 0.3 & f < 0.5;
                offset = round(mean(phaseTrans(highSnrIndices)...
                    - noDispPhaseTrans(highSnrIndices)) / pi) * pi;
                phaseTrans = phaseTrans - offset;
            end
        end
        
        % Calculate the time delay induced by the sample
        function delay = calTimeDelay(obj)
            [~, refPeakIndex] = max(obj.refX);
            refPeakTime = obj.refT(refPeakIndex);
            [~, sampPeakIndex] = max(obj.sampX);
            sampPeakTime = obj.sampT(sampPeakIndex);
            delay = sampPeakTime - refPeakTime;
        end
    end
end